### ISSUE TITLE
IP level client policy & telemetry

### ISSUE THREAD

**Issue by kyessenov** [istio/istio#41527]
This is a feature request to backport ambient extensions to apply P&T per destination IP onto the sidecar model. This is a generalization of the existing poorly documented autoMTLS and endpoint metadata labels sent through EDS, but extended to support EDS-less addressing.

There are three main components:

* Proxy workload metadata listener filter consumes a relation <destination IP, transport socket, destination telemetry metadata>, where the transport socket is one of {raw, TLS, TLS + CONNECT}, telemetry metadata is the existing peer metadata, a combination of namespace, canonical service/version, workload.

* xDS server synchronizes the proxy state for the table. For small clusters, a SotW via ECDS is sufficient. For large clusters, ODS with fine grains is preferred.

* Proxy extensions select upstream, verify SAN, propagate metadata, and report telemetry by looking up the relation.

**Comment by howardjohn** [comment]
Is this an abstract idea right now or do you have some thoughts on the implementation? Either way sounds great, if there are some more ideas would love to review them so I understand more

**Comment by kyessenov** [comment]
This is just writing an Envoy client for the workload discovery service in #41529, and then changing how policy is applied:
* automTLS is determined without needing VIP, this solves ORIGINAL_DST use case for auto mTLS;
* SAN verification is decoupled from VIP as well, policy can trust this SAN instead of mTLS (per @costinm 's trusted network setup);
* telemetry does not need on the wire info.

This has been long discussed as MetadataDiscoveryService. None of this depends on ambient or HBONE, hence, I'm referring to it as "backporting" bits and pieces to sidecars.

**Comment by howardjohn** [comment]
That makes sense, the part I wasn't sure how it would work is how the transport socket selection works. Would we just select "mTLS" and look that up somewhere, or each Pod IP has its own tls context associated, or something else?

**Comment by kyessenov** [comment]
Downstream determines the pool. So we can select a transport socket based on the options programmed by downstream. There are many ways to represent this in xDS, either some combo transport socket or extend the existing endpoint-label match to downstream metadata match.

**Comment by costinm** [comment]
Is this for ambient or sidecars ? For the first - would be great to have a design doc, probably many implications. Proxyless gRPC is an example.
For sidecar - I would minimize any change.

**Comment by kyessenov** [comment]
This is for sidecars. The issue is that we are already doing control plane xDS changes for ambient that sidecar can subscribe to and get obvious benefits. I agree that we should keep this optional for sidecars, fallbacks should continue as before.

**Comment by costinm** [comment]
I would focus on ambient - keeping sidecars stable is probably more useful at this point. 

For ambient - I personally think there is a lot of value in having the metadata sent by the XDS server and to avoid the in-line exchange and 'Baggage'. gRPC team also expressed concerns with the cost of in-line metadata ( they are very focused on 
performance ), but we will need to have some design doc on how this would work that they can review and implement as well.
We are very actively looking at how to improve telemetry for proxyless gRPC users.

**Comment by kyessenov** [comment]
@costinm I think you're saying the same thing, baggage is too expensive, and metadata service once available makes things more efficient for every mesh. Sidecars are stable, we would just provide yet another fallback to get metadata instead of 4 or 5 existing ways present. The point is that this service needs to be shared with ambient, gRPC, and as proposed here, sidecars.

**Comment by costinm** [comment]
I don't think for ambient it should be fallback - but built into the core
protocol, no Baggage if both ends are in the mesh and support this.

I'm not against using this in envoy as well - my main concern was to reduce
churn, test and design this for ambient and see how it
performs - and when we are happy and ready to freeze it - we can get it
implemented in envoy, proxyless, etc.

I mentioned in a few reviews that I think policy and telemetry have some
differences in trust requirements - and we should
use the info in signed certs or from the XDS server over untrusted in-line
meta whenever possible.

**Comment by kyessenov** [comment]
The great thing is that this can be bolted on without much damage. It'll just make things faster by removing the need to parse on the wire and provide better telemetry. Sidecars will have to support headers for who-knows-how-long for backwards compatibility. I was thinking to switch to baggage encoding instead of protobuf for header becaue it's 10x faster to decode, but I'm worried about compatibility.

Yes, cert is preferred, but it only gives SA and namespace.

**Comment by costinm** [comment]
Agreed, I think this is one of the most efficient and scalable ways to deal
with telemetry metadata, and will provide far more trust-worthy information
( from Istiod), better caching, etc. Compatibility is an issue, but only
for ambient, we can keep Sidecar as it is to minimize risk ( may need to
use
this mechanism, but only when talking with ztunnel ).

For cert - I mentioned in a PR, but at least when Citadel is used we can
encode more info in the cert, and we can start recommending other
CA integrators to do the same. Canonical service or even the list of
services associated with the pod can be included ( and updated if they
change,
we have push support and custom code now).

But even SA and namespace of the caller are a great  start - canonical
service name and SA are both good ways to identify the caller.
I know a very large org using the caller SA in most of the telemetry.

**Comment by kyessenov** [comment]
For async workload telemetry metadata, the plan is to enable this capability on sidecars:
1) Clients may block to resolve IP based policy, retrieve automTLS status, network policy, and destination metadata.
2) Servers will subscribe to the client metadata out of band. Servers will await to publish telemetry until it has the information.

xDS servers need to maintain client subscriptions, and push updated IP->metadata with reasonable latency. 

The following features can be safely deprecated after this lands:
* opportunistic metadata in EDS/CDS. That will separate the load balancer concerns from the telemetry needs (cc @costinm ), replaced by (1). 
* MX filters. That will reduce Istio overhead (less bytes in a request, less CPU spent decoding), replaced by (2).

**Comment by kyessenov** [comment]
Yes, we need a shared metadata service. Proto can be copied. It's just 4 or 5 fields, so we can always fix it up as long as it's based on on-demand delta xDS.
